### PR TITLE
metrics: chunked responses

### PR DESCRIPTION
**Pull request by koivunej**
Metrics can get really large in the order of hundreds of megabytes, which we currently buffer completly (after a few rounds of growing the buffer).

### CODE REVIEW

File: libs/utils/src/http/endpoint.rs [hunk 1]
```diff
@@ -148,27 +147,176 @@ impl Drop for RequestCancelled {
...
+
+                    tracing::info!(
+                        out_of_space,
+                        buf.len = buf.len(),
+                        pre_len,
+                        pre_cap,
+                        post_len = self.buffer.len(),
+                        post_cap = self.buffer.capacity(),
+                        "flushing"
+                    );
```
Comment (koivunej) [inline libs/utils/src/http/endpoint.rs:259]: I don't think the buffer is being reused at all, but ... we can fix that later. It will probably need some more signalling, or hacking together a bufferless channel.

File: libs/utils/src/http/endpoint.rs [hunk 2]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+
+            fn flush0(&mut self) -> std::io::Result<usize> {
+                let n = self.buffer.len();
+                if n > 0 {
+                    let ready = self.buffer.split().freeze();
+                    tracing::trace!(n = ready.len(), "flushing");
+
+                    // not ideal to call from blocking code to block_on, but we are sure that this
+                    // operation does not spawn_blocking other tasks
+                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:217]:
```suggestion
                    // not ideal to call from blocking code to block_on, but we are sure that this
                    // operation does not spawn_blocking other tasks, so, no deadlock risk
                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
```

File: libs/utils/src/http/endpoint.rs [hunk 3]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+                        self.tx.send(Ok(ready)).await.map_err(|_| ())?;
+
+                        // throttle sending to allow reuse of our buffer in `write`.
+                        self.tx.reserve().await.map_err(|_| ())?;
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:221]: This hard-depends on the channel having capacity 1.

I suggest to move channel creation into the `new()` function and return the writer and the stream as a tuple from there.

(That would also be a good opportunity to move this struct and impl into `utils`, it seems generic and useful).

File: libs/utils/src/http/endpoint.rs [hunk 4]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+
+                        // now the response task has picked up the buffer and hopefully started
+                        // sending it to the client.
+                        Ok(())
+                    });
+                    if res.is_err() {
+                        return Err(std::io::ErrorKind::BrokenPipe.into());
+                    }
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:229]: I can live with this, but why not a custom error telling that the receiver is gone?

If this error gets printed anywhere without enough context, people are going to start wondering "where are we even using a pipe".

File: libs/utils/src/http/endpoint.rs [hunk 5]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+        impl std::io::Write for ChannelWriter {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                let remaining = self.buffer.capacity() - self.buffer.len();
+
+                let out_of_space = remaining < buf.len();
+
+                if out_of_space {
+                    self.flush0()?;
+                }
+
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:249]: Would be nice to assert that we now have enough space after the flush0.

File: libs/utils/src/http/endpoint.rs [hunk 6]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+                self.buffer.extend_from_slice(buf);
+                Ok(buf.len())
+            }
+
+            fn flush(&mut self) -> std::io::Result<()> {
+                self.flush0().map(|_| ())
+            }
+        }
+
+        let mut writer = ChannelWriter::new(128 * 1024, tx);
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:262]: rx is created super far above. See my earlier request to move this struct into `utils`

File: libs/utils/src/http/endpoint.rs [hunk 7]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+                .encode(&metrics, &mut writer)
+                .and_then(|_| writer.flush().map_err(|e| e.into()));
+
+            match res {
+                Ok(()) => {
+                    tracing::info!(
+                        bytes = writer.flushed_bytes(),
+                        elapsed_ms = started_at.elapsed().as_millis(),
+                        "responded /metrics"
+                    );
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:289]: This shouldn't be `info`, no?

File: libs/utils/src/http/endpoint.rs [hunk 8]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+        let span = info_span!("blocking");
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let res = encoder
+                .encode(&metrics, &mut writer)
+                .and_then(|_| writer.flush().map_err(|e| e.into()));
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:281]: Would be helpful to annotate the `e.into()`. Reading top down, especially with the elaborate `Err(e)` below, it's hard to follow what error type this is.

I _think_ we're converting encoder error into io error? https://docs.rs/prometheus/latest/prometheus/enum.Error.html

File: libs/utils/src/http/endpoint.rs [hunk 9]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+                    tracing::warn!("failed to write out /metrics response: {e:#}");
+                    // semantics of this error are quite... unclear. we want to error the stream out to
+                    // abort the response to somehow notify the client that we failed.
+                    //
+                    // though, most likely the reason for failure is that the receiver is already gone.
+                    drop(
+                        writer
+                            .tx
+                            .blocking_send(Err(std::io::ErrorKind::BrokenPipe.into())),
+                    );
```
Comment (problame) [inline libs/utils/src/http/endpoint.rs:301]: Who's going to receive this `writer.tx.blocking_send()?`

File: libs/utils/src/http/endpoint.rs (hunk 2 above)
Comment (problame) [reply libs/utils/src/http/endpoint.rs:217]: Also, do I understand correctly that we use the spawn_blocking thread to block_on the future?

But, further down, we're doing `blocking_send` into the channel if we have a writer error. How does this not deadlock? Because when we're doing the blocking_send, we know we're not writing anymore?

**Review by problame — APPROVED** [review]
Surprisingly complicated for such trivial functionality.

I'm not sure I totally follow. Left some comments. Approving because it's off-by-default. But, we should move that struct into a `utils` lib, maybe even with an explanatory test case that showcases error handling.

File: libs/utils/src/http/endpoint.rs (hunk 8 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:281]: Yeah this was hugely frustrating. I hope we could get rid of prometheus asap.

File: libs/utils/src/http/endpoint.rs (hunk 9 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:301]: The body, who else?

It's in the off-chance that the TextEncoder fails for some other reason than i/o related.

File: libs/utils/src/http/endpoint.rs (hunk 5 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:249]: BytesMut grows, no need.

File: libs/utils/src/http/endpoint.rs (hunk 4 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:229]: This is called within the `std::io::Write` impl. Very self-contained, not used from anywhere else than this one response writing out. It doesn't matter what the error is, but these plain `ErrorKind.into()` are the cheapest ones.

File: libs/utils/src/http/endpoint.rs (hunk 2 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:217]: This is needed by a `std::io::Write` impl. It is used within the same `spawn_blocking` context as the `blocking_send`, on the same value of `tx` as used here.

So overall this works like:

`Body::wrap_stream(ReceiverStream(rx))` is being fed `std::io::Result<Bytes>` by the `std::io::Write` impl (`ChannelWriter`). The http side of this (stream) is async. The blocking side of this is the prometheus text generation (spawn_blocking).

Do you have a different idea on how to connect the two?

Also I don't understand the deadlock idea at all.

> How does this not deadlock? Because when we're doing the blocking_send, we know we're not writing anymore?

This is ensured by `&mut` which can be thought of as unique borrows here, the slice:

```rust
let mut writer = ChannelWriter::new(128 * 1024, tx);
let encoder = TextEncoder::new();
let res = encoder.encode(&metrics, &mut writer)....;
// the above call produces N of
// block_on(async {
//   self.tx.send(Ok(ready)).await...?;
//   self.tx.reserve().await...?;
// });

// in the Err path:
drop(
    writer
        .tx
        .blocking_send(Err(std::io::ErrorKind::BrokenPipe.into())),
);
```

The `tx` could be a cloned since it's mpsc, but we are just using it as spsc here, because there's no need, and only one producer.

File: libs/utils/src/http/endpoint.rs (hunk 3 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:221]: How does it hard depend on this?

> good opportunity to move

I don't want to expose this in any other place.

File: libs/utils/src/http/endpoint.rs (hunk 7 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:289]: Do you have any rational for either way?

I have been thinking that we have zero clue how long these things take. Do you know? The way we can get the precise bytes sent is by watching a counter on grafana. This is also the only line logged for this request when there is no error.

File: libs/utils/src/http/endpoint.rs [hunk 10]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+                ChannelWriter {
+                    buffer: BytesMut::with_capacity(buf_len),
+                    tx,
+                    written: 0,
+                }
+            }
+
+            fn flush0(&mut self) -> std::io::Result<usize> {
+                let n = self.buffer.len();
+                if n > 0 {
```
Comment (arpad-m) [inline libs/utils/src/http/endpoint.rs:204]: This could be an early return, you save an indentation level.

File: libs/utils/src/http/endpoint.rs (hunk 6 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:262]: I could move those lines closer to this. The types does not need to be exposed.

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:259]: Moved in 2b4a3e4ef

File: libs/utils/src/http/endpoint.rs [hunk 11]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+                Ok(()) => {
+                    tracing::info!(
+                        bytes = writer.flushed_bytes(),
+                        elapsed_ms = started_at.elapsed().as_millis(),
+                        "responded /metrics"
+                    );
+                }
+                Err(e) => {
+                    tracing::warn!("failed to write out /metrics response: {e:#}");
+                    // semantics of this error are quite... unclear. we want to error the stream out to
```
Comment (arpad-m) [inline libs/utils/src/http/endpoint.rs:287]: For the non-tokio mpsc, it's documented [more explicitly](https://doc.rust-lang.org/nightly/std/sync/mpsc/struct.SendError.html), but I think for tokio, too, it only gives an error if the receiver has been closed. Not 100% sure tho, haven't dug deep enough into tokio's code. https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#errors

File: libs/utils/src/http/endpoint.rs (hunk 10 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:204]: Yeah, it grew a bit, can do.

File: libs/utils/src/http/endpoint.rs [hunk 12]
```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
...
+    use tokio::sync::mpsc;
+    use tokio_stream::wrappers::ReceiverStream;
+
     SERVE_METRICS_COUNT.inc();
 
-    let mut buffer = vec![];
-    let encoder = TextEncoder::new();
+    // we need to first test this out with vmagent on staging, if it works, we can get turn this to
+    // true or maybe get rid of the fully buffering path altogether
+    const CHUNKED_IS_DEFAULT: bool = false;
```
[the remark this answers is not in the document]
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:159]: I'm good with that idea. Removing the rest.

File: libs/utils/src/http/endpoint.rs [hunk 13]
```diff
@@ -148,26 +147,129 @@ impl Drop for RequestCancelled {
...
+
+    impl std::io::Write for ChannelWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let remaining = self.buffer.capacity() - self.buffer.len();
+
+            let out_of_space = remaining < buf.len();
+
+            if out_of_space {
+                self.flush0()?;
+            }
```
Comment (koivunej) [inline libs/utils/src/http/endpoint.rs:214]: This is still a real problem. It might cause a lot of allocations. It would be helpful to ... Only do the half of the buffer at a time.

Perhaps I figured it just out; no code changes here, but split the buffer at half *after* allocating it. Then it will grow up to or near the "half", and extend_from_slice will be taking up the rest of the allocation.

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:214]: Well, caveats remain even with this. There are cases where the prometheus does do small writes, but perhaps we can asusme this is fine.

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:214]: At least with fast localhost connection this works, but probably breaks down with slower.

```
2023-07-21T13:14:55.435907Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65563
2023-07-21T13:14:55.436134Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65482
2023-07-21T13:14:55.436321Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65491
2023-07-21T13:14:55.436504Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65535
2023-07-21T13:14:55.436783Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65508
2023-07-21T13:14:55.436963Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65530
2023-07-21T13:14:55.437146Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65541
2023-07-21T13:14:55.437348Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65528
2023-07-21T13:14:55.437537Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65536
2023-07-21T13:14:55.437786Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=57983
2023-07-21T13:14:55.437816Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: responded /metrics bytes=647697 elapsed_ms=2
```

Assuming good enough for now.

File: libs/utils/src/http/endpoint.rs (hunk 1 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: Later thought of splitting the BytesMut right after allocating.

File: libs/utils/src/http/endpoint.rs (hunk 3 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: Not a hard dependency. If you put (in the same function) a huge number as capacity, it might turn into like sparsely allocating all of the body depending on the slowloris client.

File: libs/utils/src/http/endpoint.rs (hunk 8 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: This transform is made *because* TextEncoder can fail for non-IO reasons, which we handle in the most confusing `blocking_send(Err(BrokenPipe.into()))` for which I don't know a better alternative (we want to signal to hyper that something went bad, drop the connection).

File: libs/utils/src/http/endpoint.rs (hunk 10 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: In 88a2cda54a98d7a523d9b8759f0bbf91c2f2f51f

File: libs/utils/src/http/endpoint.rs (hunk 11 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: Since it's an `mpsc` multiple producer single consumer it will error out. With this confusing comment I was after the semantics of what *hyper* will do upon seeing an Error. I recall it will do nothing, maybe it now closes the connection.

But it matters very little for `/metrics`. We are already saving a lot of allocations with this.

File: libs/utils/src/http/endpoint.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: [cb76bb0](https://github.com/neondatabase/neon/pull/4768/commits/cb76bb0298670207e49c67bdf43557047fb98098) and fixup [ada0dd9](https://github.com/neondatabase/neon/pull/4768/commits/ada0dd987c6d59c36f7414e9009cc93890a2c8ce)

File: libs/utils/src/http/endpoint.rs (hunk 13 above)
Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: (I forgot to return the full length from the write.)

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs]: [1ccf599](https://github.com/neondatabase/neon/pull/4768/commits/1ccf599548937ca8d1e8784074c718a1d2034d29)
